File: 977/977.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/977.py

val MODVAL: Long = 1_000_000_007L

fun modpow(base: Long, exp: Long, modulus: Long): Long {
    var result = 1L
    var b = base % modulus
    var e = exp
    while (e > 0L) {
        if (e % 2L == 1L) {
            result = result * b % modulus
        }
        e /= 2L
        b = b * b % modulus
    }
    return result
}

fun computeMod(n: Int): Long {
    val N = n.toLong()
    if (N == 1L) return 1L
    var total = 0L
    val mVal = N - 2L
    val sumQ = (mVal * (mVal + 1L) * (2L * mVal + 1L) / 6L + mVal * (mVal + 1L) / 2L) % MODVAL
    total = (sumQ + N) % MODVAL
    for (L in 2L..N) {
        val R = N - L
        if (R >= 1L) {
            val qFull = (R - 1L) / L
            val maxA = qFull + 2L
            val powA = LongArray((maxA + 1).toInt())
            if (L == 2L) {
                for (a in 1L..maxA) {
                    powA[a.toInt()] = (a * a) % MODVAL
                }
            } else if (L == 3L) {
                for (a in 1L..maxA) {
                    val aa = a * a % MODVAL
                    powA[a.toInt()] = aa * a % MODVAL
                }
            } else {
                for (a in 1L..maxA) {
                    powA[a.toInt()] = modpow(a, L, MODVAL)
                }
            }
            for (q in 0L until qFull) {
                val A = q + 1L
                val B = q + 2L
                val AL = powA[A.toInt()]
                val BL = powA[B.toInt()]
                val AL1 = AL * A % MODVAL
                val term = ((q * AL + (A * A % MODVAL) * BL - B * AL1) % MODVAL + MODVAL) % MODVAL
                total = (total + term) % MODVAL
            }
            run {
                val q = qFull
                val mInner = (R - 1L) - qFull * L
                val A = q + 1L
                val B = q + 2L
                val AL = powA[A.toInt()]
                var term = q * AL % MODVAL
                if (mInner >= 1L) {
                    val AL1 = AL * A % MODVAL
                    val expVal = L + 1L - mInner
                    val AL1m: Long = when (expVal) {
                        1L -> A % MODVAL
                        2L -> A * A % MODVAL
                        3L -> A * A % MODVAL * A % MODVAL
                        else -> modpow(A, expVal, MODVAL)
                    }
                    val Bm: Long = when (mInner) {
                        1L -> B % MODVAL
                        2L -> B * B % MODVAL
                        3L -> B * B % MODVAL * B % MODVAL
                        else -> modpow(B, mInner, MODVAL)
                    }
                    term = (term + B * ((AL1m * Bm - AL1) % MODVAL + MODVAL) % MODVAL) % MODVAL
                }
                total = (total + term) % MODVAL
            }
        }
        val q = R / L
        val r = R - q * L
        val A = q + 1L
        val B = q + 2L
        val base: Long = if (r == 0L) {
            modpow(A, L, MODVAL)
        } else {
            modpow(A, L - r, MODVAL) * modpow(B, r, MODVAL) % MODVAL
        }
        total = (total + base) % MODVAL
    }
    return total % MODVAL
}

fun computeExact(n: Int): java.math.BigInteger {
    val N = java.math.BigInteger.valueOf(n.toLong())
    val ONE = java.math.BigInteger.ONE
    val TWO = java.math.BigInteger.TWO
    val THREE = java.math.BigInteger.valueOf(3L)
    val SIX = java.math.BigInteger.valueOf(6L)
    if (n == 1) return ONE
    var total = java.math.BigInteger.ZERO
    val mVal = N - TWO
    val sumQ = mVal * (mVal + ONE) * (TWO * mVal + ONE) / SIX + mVal * (mVal + ONE) / TWO
    total = sumQ + N
    var L = TWO
    while (L <= N) {
        val R = N - L
        if (R >= ONE) {
            val qFull = (R - ONE) / L
            val maxA = qFull + TWO
            val powA = arrayOfNulls<java.math.BigInteger>((maxA.toLong() + 1L).toInt())
            var a = ONE
            while (a <= maxA) {
                powA[a.toInt()] = a.pow(L.toInt())
                a += ONE
            }
            var q = java.math.BigInteger.ZERO
            while (q < qFull) {
                val A = q + ONE
                val B = q + TWO
                val AL = powA[A.toInt()]!!
                val BL = powA[B.toInt()]!!
                val term = q * AL + (A * A) * BL - B * (AL * A)
                total += term
                q += ONE
            }
            run {
                val q2 = qFull
                val mInner = (R - ONE) - qFull * L
                val A = q2 + ONE
                val B = q2 + TWO
                val AL = powA[A.toInt()]!!
                var term = q2 * AL
                if (mInner >= ONE) {
                    val AL1 = AL * A
                    val AL1m = A.pow((L + ONE - mInner).toInt())
                    val Bm = B.pow(mInner.toInt())
                    term += B * (AL1m * Bm - AL1)
                }
                total += term
            }
        }
        val q = R / L
        val r = R - q * L
        val A = q + ONE
        val B = q + TWO
        val base = A.pow((L - r).toInt()) * B.pow(r.toInt())
        total += base
        L += ONE
    }
    return total
}

fun main() {
    check(computeExact(3) == java.math.BigInteger.valueOf(8L)) { "assert failed: computeExact(3)" }
    check(computeExact(7) == java.math.BigInteger.valueOf(174L)) { "assert failed: computeExact(7)" }
    check(computeExact(100) == java.math.BigInteger("570271270297640131")) { "assert failed: computeExact(100)" }
    val n = 1_000_000
    println(computeMod(n))
}

File: 977/977.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/977.py

const MODVAL: u64 = 1_000_000_007;

fn modpow(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    let mut result: u64 = 1;
    base %= modulus;
    while exp > 0 {
        if exp % 2 == 1 {
            result = ((result as u128 * base as u128) % modulus as u128) as u64;
        }
        exp /= 2;
        base = ((base as u128 * base as u128) % modulus as u128) as u64;
    }
    result
}

fn compute_mod(n: i64) -> u64 {
    if n == 1 {
        return 1;
    }
    let mut total: u64 = 0;
    let m_val = n - 2;
    let sum_q = (m_val * (m_val + 1) * (2 * m_val + 1) / 6 + m_val * (m_val + 1) / 2) as u64 % MODVAL;
    total = (sum_q + n as u64) % MODVAL;
    for l_val in 2i64..=n {
        let r_val = n - l_val;
        if r_val >= 1 {
            let q_full = (r_val - 1) / l_val;
            let max_a = q_full + 2;
            let mut pow_a: Vec<u64> = vec![0u64; (max_a + 1) as usize];
            if l_val == 2 {
                for a in 1i64..=max_a {
                    pow_a[a as usize] = (a as u64 * a as u64) % MODVAL;
                }
            } else if l_val == 3 {
                for a in 1i64..=max_a {
                    let aa = (a as u64 * a as u64) % MODVAL;
                    pow_a[a as usize] = (aa * a as u64) % MODVAL;
                }
            } else {
                for a in 1i64..=max_a {
                    pow_a[a as usize] = modpow(a as u64, l_val as u64, MODVAL);
                }
            }
            for q in 0i64..q_full {
                let a_val = q + 1;
                let b_val = q + 2;
                let al = pow_a[a_val as usize];
                let bl = pow_a[b_val as usize];
                let al1 = ((al as u128 * a_val as u128) % MODVAL as u128) as u64;
                let t1 = ((q as u128 * al as u128) % MODVAL as u128) as u64;
                let t2 = (((a_val as u128 * a_val as u128) % MODVAL as u128) * bl as u128 % MODVAL as u128) as u64;
                let t3 = ((b_val as u128 * al1 as u128) % MODVAL as u128) as u64;
                let term = (t1 + t2 + MODVAL - t3) % MODVAL;
                total = (total + term) % MODVAL;
            }
            {
                let q = q_full;
                let m_inner = (r_val - 1) - q_full * l_val;
                let a_val = q + 1;
                let b_val = q + 2;
                let al = pow_a[a_val as usize];
                let mut term = ((q as u128 * al as u128) % MODVAL as u128) as u64;
                if m_inner >= 1 {
                    let al1 = ((al as u128 * a_val as u128) % MODVAL as u128) as u64;
                    let exp_val = l_val + 1 - m_inner;
                    let al1m = match exp_val {
                        1 => a_val as u64 % MODVAL,
                        2 => ((a_val as u128 * a_val as u128) % MODVAL as u128) as u64,
                        3 => {
                            let tmp = ((a_val as u128 * a_val as u128) % MODVAL as u128) as u64;
                            ((tmp as u128 * a_val as u128) % MODVAL as u128) as u64
                        }
                        _ => modpow(a_val as u64, exp_val as u64, MODVAL),
                    };
                    let bm = match m_inner {
                        1 => b_val as u64 % MODVAL,
                        2 => ((b_val as u128 * b_val as u128) % MODVAL as u128) as u64,
                        3 => {
                            let tmp = ((b_val as u128 * b_val as u128) % MODVAL as u128) as u64;
                            ((tmp as u128 * b_val as u128) % MODVAL as u128) as u64
                        }
                        _ => modpow(b_val as u64, m_inner as u64, MODVAL),
                    };
                    let inner = (((al1m as u128 * bm as u128) % MODVAL as u128) as u64 + MODVAL - al1) % MODVAL;
                    let add = ((b_val as u128 * inner as u128) % MODVAL as u128) as u64;
                    term = (term + add) % MODVAL;
                }
                total = (total + term) % MODVAL;
            }
        }
        let q = r_val / l_val;
        let r = r_val - q * l_val;
        let a_val = q + 1;
        let b_val = q + 2;
        let base_val = if r == 0 {
            modpow(a_val as u64, l_val as u64, MODVAL)
        } else {
            let pa = modpow(a_val as u64, (l_val - r) as u64, MODVAL);
            let pb = modpow(b_val as u64, r as u64, MODVAL);
            ((pa as u128 * pb as u128) % MODVAL as u128) as u64
        };
        total = (total + base_val) % MODVAL;
    }
    total % MODVAL
}

fn compute_exact(n: i64) -> i128 {
    if n == 1 {
        return 1;
    }
    let mut total: i128 = 0;
    let m_val = n - 2;
    let sum_q = (m_val * (m_val + 1) * (2 * m_val + 1) / 6 + m_val * (m_val + 1) / 2) as i128;
    total = sum_q + n as i128;
    for l_val in 2i64..=n {
        let r_val = n - l_val;
        if r_val >= 1 {
            let q_full = (r_val - 1) / l_val;
            let max_a = q_full + 2;
            let mut pow_a: Vec<i128> = vec![0i128; (max_a + 1) as usize];
            for a in 1i64..=max_a {
                pow_a[a as usize] = (a as i128).pow(l_val as u32);
            }
            for q in 0i64..q_full {
                let a_val = q + 1;
                let b_val = q + 2;
                let al = pow_a[a_val as usize];
                let bl = pow_a[b_val as usize];
                let term = q as i128 * al + (a_val as i128 * a_val as i128) * bl - b_val as i128 * (al * a_val as i128);
                total += term;
            }
            {
                let q = q_full;
                let m_inner = (r_val - 1) - q_full * l_val;
                let a_val = q + 1;
                let b_val = q + 2;
                let al = pow_a[a_val as usize];
                let mut term = q as i128 * al;
                if m_inner >= 1 {
                    let al1 = al * a_val as i128;
                    let al1m = (a_val as i128).pow((l_val + 1 - m_inner) as u32);
                    let bm = (b_val as i128).pow(m_inner as u32);
                    term += b_val as i128 * (al1m * bm - al1);
                }
                total += term;
            }
        }
        let q = r_val / l_val;
        let r = r_val - q * l_val;
        let a_val = q + 1;
        let b_val = q + 2;
        let base_val = (a_val as i128).pow((l_val - r) as u32) * (b_val as i128).pow(r as u32);
        total += base_val;
    }
    total
}

fn main() {
    assert_eq!(compute_exact(3), 8, "assert failed: compute_exact(3)");
    assert_eq!(compute_exact(7), 174, "assert failed: compute_exact(7)");
    assert_eq!(compute_exact(100), 570271270297640131, "assert failed: compute_exact(100)");
    let n = 1_000_000i64;
    println!("{}", compute_mod(n));
}
